internal/pkg/service/common/etcdop/op/atomic_core.go

```diff
@@ -2,112 +2,217 @@ package op
+//
+// By default, the feature is enabled and checks that each loaded key within the Read Phase, from a prefix, exists in Write Phase.
+// This can be potentially SLOW and generate a lot of IF conditions, if there are a large number of keys in the prefix.
+// Therefore, this feature can be turned off by the method.
+//
+// Modification of a key in the prefix is always detected,
+// this feature is used to detect the deletion of a key from the prefix.
+//
+// See TestAtomicOp:GetPrefix_DeleteKey_SkipPrefixKeysCheck.
+func (v *AtomicOpCore) SkipPrefixKeysCheck() *AtomicOpCore {
+	v.checkPrefixKey = false
 	return v
 }
 
-func (v *AtomicOpCore) RequireLock(lock mutex) *AtomicOpCore {
-	v.locks = append(v.locks, lock)
+// Read adds operations factories to the READ phase.
+//
+// The factory can return <nil>, if you want to execute some code during the READ phase,
+// but no etcd operation is generated.
+//
+// The factory can return op.ErrorOp(err) OR op.ErrorTxn[T](err) to signal a static error.
+func (v *AtomicOpCore) Read(factories ...HighLevelFactory) *AtomicOpCore {
+	v.readPhase = append(v.readPhase, factories...)
 	return v
 }
 
-func (v *AtomicOpCore) ReadOp(ops ...Op) *AtomicOpCore {
-	for _, op := range ops {
-		v.Read(func(ctx context.Context) Op {
-			return op
-		})
-	}
+// Write adds operations factories to the WRITE phase.
+//
+// The factory can return <nil>, if you want to execute some code during the READ phase,
+// but no etcd operation is generated.
+//
+// The factory can return op.ErrorOp(err) OR op.ErrorTxn[T](err) to signal a static error.
+func (v *AtomicOpCore) Write(factories ...HighLevelFactory) *AtomicOpCore {
+	v.writePhase = append(v.writePhase, factories...)
 	return v
 }
 
-func (v *AtomicOpCore) Read(factories ...func(ctx context.Context) Op) *AtomicOpCore {
-	for _, fn := range factories {
-		v.ReadOrErr(func(ctx context.Context) (Op, error) {
-			return fn(ctx), nil
-		})
-	}
-	return v
+// ReadPhaseOps returns all op factories for READ phase,
+// is used in joining two atomic operations, see AddFrom method.
+func (v *AtomicOpCore) ReadPhaseOps() (out []HighLevelFactory) {
+	return slices.Clone(v.readPhase)
 }
 
-func (v *AtomicOpCore) OnRead(fns ...func(ctx context.Context)) *AtomicOpCore {
-	for _, fn := range fns {
-		v.ReadOrErr(func(ctx context.Context) (Op, error) {
-			fn(ctx)
-			return nil, nil
-		})
+// WritePhaseOps returns all op factories for WRITE phase,
+// is used in joining two atomic operations, see AddFrom method.
+func (v *AtomicOpCore) WritePhaseOps() (out []HighLevelFactory) {
+	// If there is no processor callback and no lock (IF condition),
+	// we can return operations without wrapping them to a TXN.
+	if v.processorFactory == nil && len(v.locks) == 0 {
+		return slices.Clone(v.writePhase)
 	}
+
+	return []HighLevelFactory{func(ctx context.Context) Op {
+		txn, err := v.writeTxn(ctx)
+		if err == nil {
+			return txn
+		} else {
+			return ErrorOp(err)
+		}
```

Comment:
this use of if{}else{} always confuses me, but you can ignore it and leave it

Reply:
I see no real difference in this short code:

My version:
```go
if txn, err := v.writeTxn(ctx); err == nil {
	return txn
} else {
	return ErrorOp(err)
}
```

You version:
```go
txn, err := v.writeTxn(ctx)
if err == nil {
	return ErrorOp(err)
}
return txn
```

@jachym-tousek-keboola @Matovidlo what do you think?

Reply:
txn, err := v.writeTxn(ctx)
if err != nil {
	return ErrorOp(err)
}
return txn

Reply:
(thanks, I fixed the example)

Reply:
I agree with above. conversation

Reply:
`err == nil` .... `100x` in th repo - @hosekpeter please, create task to add a linter and replace all places.